scenario/runtime.py

```diff
@@ -58,6 +54,192 @@ class DirtyVirtualCharmRootError(ScenarioRuntimeError):
     """Error raised when the runtime can't initialize the vroot without overwriting existing metadata files."""
 
 
+class InconsistentScenarioError(ScenarioRuntimeError):
+    """Error raised when the combination of state and event is inconsistent."""
+
+
+class ConsistencyChecker:
+    """This class is responsible for validating the combination of a state, an event, a charm spec, and a juju version.
+
+    Upon calling .run(), it performs a series of checks that validate that the state is consistent with itself, with
+    the event being emitted, the charm metadata, etc...
+    For example, if someone tries to emit a foo-pebble-ready but there is no container 'foo' in metadata or in State,
+    this is where we surface the issue.
+    """
+
+    def __init__(
+        self,
+        state: "State",
+        event: "Event",
+        charm_spec: "_CharmSpec",
+        juju_version: str,
+    ):
+        self.state = state
+        self.event = event
+        self.charm_spec = charm_spec
+        self.juju_version: Tuple[int, ...] = tuple(map(int, juju_version.split(".")))
+
+    def run(self):
+        """Run all consistency checks and raise if any of them fails."""
+        if os.getenv("SCENARIO_SKIP_CONSISTENCY_CHECKS"):
+            logger.info("skipping consistency checks.")
+            return
+
+        errors = []
+
+        for check in (
+            self._check_containers,
+            self._check_config,
+            self._check_event,
+            self._check_secrets,
+        ):
+            try:
+                results = check()
+            except Exception as e:
```

Comment:
Do we really want to catch all exceptions here? This will also "catch" any bugs in our checker code. It seems to me like if there is a bug in the checker that *should* raise an exception here, not return a validation error.